Introducing "took" time (in ms) for _msearch #23131

olcbean opened the pull request:
As per https://github.com/elastic/elasticsearch/issues/23131#issuecomment-280307124 it is not clear if  introducing took time for multi search is a desired behavior or if a further discussion is needed.

But as there are no comments against it so far I am submitting this PR.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

olcbean commented:
Hm...
I signed the CLA, but still the check is failing...

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/TransportMultiSearchActionTests.java:
```diff
@@ -63,15 +63,18 @@ public void testBatchExecute() throws Exception {
+        TransportService transportService = new TransportService(Settings.EMPTY, null, null,
+                TransportService.NOOP_TRANSPORT_INTERCEPTOR,
+                boundAddress -> DiscoveryNode.createLocal(settings, boundAddress.publishAddress(),
+                        UUIDs.randomBase64UUID()), null) {
```
Please revert all line-length changes that are changing lines that otherwise were not impacted by the changes in this PR.

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/TransportMultiSearchActionTests.java:
```diff
@@ -102,14 +111,17 @@ protected void doExecute(SearchRequest request, ActionListener<SearchResponse> l
+
+        TransportMultiSearchAction action = new TransportMultiSearchAction(threadPool,
+                actionFilters, transportService, clusterService, searchAction, resolver, 10,
+                System::nanoTime);
```
As far as I can tell, this is the only change in this file that was necessitated by the changes in this PR.

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java:
```diff
@@ -0,0 +1,221 @@
+        clusterService.close();
+    }
+    
+    private enum Clock {
```
Please remove, a simple boolean like in the other took tests is fine.

olcbean replied:
It was added to readability but I will replace it with a boolean.

olcbean commented:
@jasontedor I hope I addressed all of your remarks.
Thank you for your review :)

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java:
```diff
@@ -111,11 +112,14 @@ public Exception getFailure() {
     }
 
-    public MultiSearchResponse(Item[] items) {
+    public MultiSearchResponse(long tookInMillis, Item[] items) {
```
Let's flip the order of these parameters (in general, I think that new parameters should be added at the end absent a good reason otherwise, but putting it at the end also aligns it more closely with the `BulkResponse`; it wouldn't surprise me we are not consistent about this through the codebase, but I'm choosing the `BulkResponse` here to be consistent with).

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java:
```diff
@@ -111,11 +112,14 @@ public Exception getFailure() {
 
-    public MultiSearchResponse(Item[] items) {
+    public MultiSearchResponse(long tookInMillis, Item[] items) {
+        this.tookInMillis = tookInMillis;
```
Let's move this parameter down after the order of the parameter is changed.

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java:
```diff
@@ -111,11 +112,14 @@ public Exception getFailure() {
     }
 
     private Item[] items;
+    
```
Can you remove this unnecessary space?

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java:
```diff
@@ -130,6 +134,20 @@ public MultiSearchResponse(Item[] items) {
+    /**
+     * How long the msearch took.
+     */
+    public TimeValue getTook() {
```
It looks like this is only used in tests, and only to get the milliseconds, so I think that we can remove it and just use `getTookInMillis` in those places.

olcbean replied:
The method exists in both `SearchResponse` and `BulkResponse`.
I would prefer to keep the consistency unless there is a good reason no to.

jasontedor replied:
There is a good reason, the method is unneeded.

olcbean replied:
So shall  remove also the corresponding method for `Bulk` and `Search`?

jasontedor replied:
Not as part of this change.

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java:
```diff
@@ -138,6 +156,7 @@ public void readFrom(StreamInput in) throws IOException {
         for (int i = 0; i < items.length; i++) {
             items[i] = Item.readItem(in);
         }
+        tookInMillis = in.readVLong();
```
There are backwards compatibility concerns here. Consider a mixed-version cluster that contains nodes that do not know about took. You need to guard this read in a version check on the stream. For now, let's try for on or after 5.4.0. We can adjust this later if needed.

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java:
```diff
@@ -147,11 +166,13 @@ public void writeTo(StreamOutput out) throws IOException {
         for (Item item : items) {
             item.writeTo(out);
         }
+        out.writeVLong(tookInMillis);
```
There is a backwards compatibility concern here too.

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -52,19 +55,23 @@ public TransportMultiSearchAction(Settings settings, ThreadPool threadPool, Tran
         this.clusterService = clusterService;
         this.searchAction = searchAction;
         this.availableProcessors = EsExecutors.numberOfProcessors(settings);
+        relativeTimeProvider = System::nanoTime;
```
Can we prefix this with `this`, it will make it symmetric with the other fields, and the other constructor.

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
In `TransportBulkAction`, the corresponding method is named `relativeTime`; let's do the same here.

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -110,10 +117,11 @@ static int defaultMaxConcurrentSearches(int availableProcessors, ClusterState st
             final AtomicArray<MultiSearchResponse.Item> responses,
             final AtomicInteger responseCounter,
+            long startTimeInNanos,
             final ActionListener<MultiSearchResponse> listener) {
```
Let's change the signature of this method to accept an `ActionListener<Item[]>`, and then invoke this method by defining a new `listener` that wraps the `ActionListener<MultiSearchResponse>` and takes care of creating the `MutliSearchResponse` with the took time. This removes the need for threading `startTimeInNanos` all the way through execute search which I think simplifies things.

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
In `TransportBulkAction`, the corresponding method is named `buildTookInMillis`; let's do the same here.

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -177,7 +197,5 @@ private void finish() {
             this.request = request;
             this.responseSlot = responseSlot;
         }
-
```
Please revert this change.

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -177,7 +197,5 @@ private void finish() {
         }
-
     }
-
```
Please revert this change.

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java:
```diff
@@ -0,0 +1,217 @@
+ */
+public class MultiSearchActionTookTests extends ESTestCase {
+
+    private static ThreadPool threadPool;
```
Let's make this an instance field and set it up and tear it down in test setup and teardown methods rather than using static. Yes, I think that this is a departure from how we do it in other places but I think doing it this way is better.

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java:
```diff
@@ -0,0 +1,217 @@
+
+    @After
+    public void tearDown() throws Exception {
+        super.tearDown();
```
The super teardown should be done after the specific teardown.

olcbean replied:
Agreed. However I was closely following the TransportBulkActionTookTests. Maybe you should change it there too?

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java:
```diff
@@ -0,0 +1,217 @@
+        clusterService.close();
+    }
+    
+    //Unit conversion using a controller clock
```
Can you add a space between the `//` and `Unit`, add the word `test` and lowercase `Unit`?

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java:
```diff
@@ -0,0 +1,217 @@
+        runTestTook(true);
+    }
+
+    //Using {@link System#nanoTime()}
```
Let's remove the use of Javadoc syntax here since this is not a Javadoc, add a space between `//` and `Using`, add the word `test`, and lowercase `Using`.

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java:
```diff
@@ -0,0 +1,217 @@
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.warn(e);
```
This should probably be a fail.

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/MultiSearchIT.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * MultiSearch IT
+ */
+public class MultiSearchIT extends ESSingleNodeTestCase {
```
I do not think that we need an integration test for this. I appreciate that you wrote it, but I think unit test suffices.

dakrone commented:
@olcbean thanks for opening this! It looks like it needs to have master merged in to fix some of the version constants, also, is it ready for another review by @jasontedor?

olcbean commented:
@dakrone thanks for following up on this :) 

I am currently traveling, will rebase at the start of next week.

olcbean commented:
@jasontedor I just rebased on top of the current master and set the bwc version to 7.0 ( perhaps should be changed to 6.0 ? ). I also added  a short note to the migration docs - feel free to leave it off if the change is not sufficiently notable.  

Could you please trigger the CI and let me know if any further changes are needed. 

CC-ing @dakrone as he also expressed interest in this PR.

dakrone commented:
@elasticmachine test this please

olcbean commented:
It has been about a month with no communication on this PR. Could you please let me know what the next steps are? @jasontedor @dakrone @rjernst

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -34,16 +35,19 @@
     private final ClusterService clusterService;
     private final TransportAction<SearchRequest, SearchResponse> searchAction;
+    private final LongSupplier relativeTimeProvider;
+    private SetOnce<Long> startTimeInNanos;
```
I don't think `SetOnce<Long>` here is desirable, that's two extra allocations on every multi-search request.

jasontedor replied:
The immutability is a positive though, we can obtain this by pushing the start time all the way down via the execute method (i.e., add it as a parameter and chain it through all the way down to finish).

olcbean replied:
@jasontedor good point! Thanks.

olcbean commented:
@jasontedor thank you for picking up this PR. Can you have another look?

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -123,7 +121,8 @@ void executeSearch(
             final AtomicInteger responseCounter,
-            final ActionListener<MultiSearchResponse> listener) {
+            final ActionListener<MultiSearchResponse> listener,
+            long startTimeInNanos) {
```
Please make this parameter final, and name it `relativeStartTime` (to make it clear that it's start time, but it's a relative and not absolute start time).

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -72,7 +70,7 @@ public TransportMultiSearchAction(Settings settings, ThreadPool threadPool, Tran
     @Override
     protected void doExecute(MultiSearchRequest request, ActionListener<MultiSearchResponse> listener) {
-        startTimeInNanos = new SetOnce<>(relativeTime());
+        long startTimeInNanos = relativeTime();
```
This can be final, and change it's name `relativeStartTime`.

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/MultiSearchActionTookTests.java:
```diff
@@ -164,9 +165,9 @@ protected void doExecute(SearchRequest request, ActionListener<SearchResponse> l
                 @Override
                 void executeSearch(final Queue<SearchRequestSlot> requests, final AtomicArray<MultiSearchResponse.Item> responses,
-                        final AtomicInteger responseCounter, final ActionListener<MultiSearchResponse> listener) {
+                        final AtomicInteger responseCounter, final ActionListener<MultiSearchResponse> listener, long startTimeInNanos) {
```
Same suggestion about the addition parameter.

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
I don't think this method is needed.

olcbean commented:
@jasontedor can you have another look?

jasontedor commented:
test this please

olcbean commented:
@jasontedor so how does it look?

jasontedor commented on docs/reference/migration/migrate_7_0/java.asciidoc:
```diff
@@ -0,0 +1,9 @@
+[[breaking_70_java_changes]]
```
We typically do not add changes here for features adds, preferring to reserve it for changes users need to know to help them migrate; I see this as a feature add. Could you please remove this?

olcbean commented:
@jasontedor just pushed the latest changes. Take another look?

jasontedor commented:
test this please

jasontedor commented:
test this please

jasontedor commented:
Thanks @olcbean.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?
